### PR TITLE
Fix VectorException test to actually test OCall

### PR DESCRIPTION
**Pull request by jhand2**
The VectorException test has code to test making an OCall from inside the exception handler but that test case is never called. Call the test case.

### CODE REVIEW

**Comment by jhand2** [comment]
bors try

**Comment by anakrish** [comment]
@jhand2 
Couple of suggestions regarding the release mode failures:
- The idiv instruction can take up 2 bytes (e.g idiv eax) or 3 bytes (e.g idiv r8) depending on which register is used. It could be possible that the compiler is using a register that triggers 3 byte instruction instead of 2 as expected by the test.
- To make ocalls work correctly, we would need to restore the host_rsp, host_rbp fields in td at the right time.
  ```c
      td->host_rbp = td->host_previous_rbp;
      td->host_rsp = td->host_previous_rsp;
      td->host_ecall_context = td->host_previous_ecall_context;
  ```
  Currently we do it after invoking the handler but before continuing the execution. Since handlers can also make ocalls, we would need to restore these fields before invoking the handler.

**Comment by jhand2** [comment]
> @jhand2
> Couple of suggestions regarding the release mode failures:
> 
>     * The idiv instruction can take up 2 bytes (e.g idiv eax) or 3 bytes (e.g idiv r8) depending on which register is used. It could be possible that the compiler is using a register that triggers 3 byte instruction instead of 2 as expected by the test.

Ah I didn't know that. Thanks for the heads up :smile:

> 
>     * To make ocalls work correctly, we would need to restore the host_rsp, host_rbp fields in td at the right time.
>       ```c
>           td->host_rbp = td->host_previous_rbp;
>           td->host_rsp = td->host_previous_rsp;
>           td->host_ecall_context = td->host_previous_ecall_context;
>       ```
>       
>       
>       Currently we do it after invoking the handler but before continuing the execution. Since handlers can also make ocalls, we would need to restore these fields before invoking the handler.

I'm not sure I understand why they need to be restored in order to make an ocall. If that was the case wouldn't all tests fail rather than just having transient failures?

**Comment by anakrish** [comment]
> I'm not sure I understand why they need to be restored in order to make an ocall. If that was the case wouldn't all tests fail rather than just having transient failures?

Only those tests that make an ocall in an exception handler would fail.

One interesting experiment to try would be to comment out the lines that restore those fields and run tests. Most tests that raise an exception (say use cpuid) and then make an ocall should fail.

**Comment by jhand2** [comment]
> > I'm not sure I understand why they need to be restored in order to make an ocall. If that was the case wouldn't all tests fail rather than just having transient failures?
> 
> Only those tests that make an ocall in an exception handler would fail.
> 
> One interesting experiment to try would be to comment out the lines that restore those fields and run tests. Most tests that raise an exception (say use cpuid) and then make an ocall should fail.

You're right, setting host rbp and rsp does fix the problem due to the explicit contract that host rsp and rbp will not be clobbered by calling enclu. Thanks for the help!

bors try

**Comment by jhand2** [comment]
@anakrish and @mikbras could you take another look at this? This should work for allowing ocalls from the enclave. Thanks!

**Comment by jhand2** [comment]
> Can you also fix the existing zero division test so that the compiler always generates 2 byte instruction?
> It is an testing issue waiting to happen.

I can't find any documentation about the size of the idiv instruction being variable. Can you point me to anything? Otherwise I'm not sure how to ensure the instruction width.

**Comment by anakrish** [comment]
> 
> 
> > Can you also fix the existing zero division test so that the compiler always generates 2 byte instruction?
> > It is an testing issue waiting to happen.
> 
> I can't find any documentation about the size of the idiv instruction being variable. Can you point me to anything? Otherwise I'm not sure how to ensure the instruction width.

You can try out the assembler here: https://defuse.ca/online-x86-assembler.htm

`idiv ecx` has machine code `f7 f9` wheras `idiv r8` has machine code `49 f7 f8`.

File: tests/VectorException/enc/enc.c [hunk 1]
```diff
@@ -285,27 +285,29 @@ int enc_test_vector_exception()
...
-    oe_exception_record_t* exception_record)
+void call_invalid_instruction()
 {
-    if (exception_record->code != OE_EXCEPTION_DIVIDE_BY_ZERO)
+    asm volatile("ud2;");
+}
+
+uint64_t test_sigill_handler_with_ocall(oe_exception_record_t* exception_record)
+{
+    if (exception_record->code != OE_EXCEPTION_ILLEGAL_INSTRUCTION)
```
Comment (anakrish) [inline tests/VectorException/enc/enc.c]: One good way to lock down the behavior would be to take a snapshot of the registers before the ocall and then after return from the ocall. The non-volatile registers ought to be preserved.

Comment (jhand2) [reply tests/VectorException/enc/enc.c]: Added to the test and modified idiv to only generate a 2 byte instruction

**Comment by jhand2** [comment]
bors try

**Comment by jhand2** [comment]
bors try-

File: tests/VectorException/enc/enc.c (hunk 1 above)
Comment (anakrish) [reply tests/VectorException/enc/enc.c]: Thanks.

**Comment by jhand2** [comment]
Fixed release tests by marking clobbered registers in inline asm

bors r+

**Comment by jhand2** [comment]
Use memory in inline asm to avoid clobbering registers before they are read on release builds.

bors r+

**Pull request merged**